bin/p2-preparer/main.go

```diff
@@ -29,24 +32,101 @@ func main() {
 	defer prep.Close()
 
 	logger.WithFields(logrus.Fields{
-		"starting":  true,
-		"node_name": preparerConfig.NodeName,
-		"consul":    preparerConfig.ConsulAddress,
-		"hooks_dir": preparerConfig.HooksDirectory,
-		"keyring":   preparerConfig.Auth["keyring"],
-		"version":   version.VERSION,
+		"starting":    true,
+		"node_name":   preparerConfig.NodeName,
+		"consul":      preparerConfig.ConsulAddress,
+		"hooks_dir":   preparerConfig.HooksDirectory,
+		"status_port": preparerConfig.StatusPort,
+		"keyring":     preparerConfig.Auth["keyring"],
+		"version":     version.VERSION,
 	}).Infoln("Preparer started successfully")
 
+	successMainUpdate := make(chan struct{})
+	successHookUpdate := make(chan struct{})
+	errMainUpdate := make(chan error)
+	errHookUpdate := make(chan error)
 	quitMainUpdate := make(chan struct{})
 	quitHookUpdate := make(chan struct{})
-	go prep.WatchForPodManifestsForNode(quitMainUpdate)
-	go prep.WatchForHooks(quitHookUpdate)
+	go prep.WatchForPodManifestsForNode(successMainUpdate, errMainUpdate, quitMainUpdate)
+	go prep.WatchForHooks(successHookUpdate, errHookUpdate, quitHookUpdate)
+
+	if preparerConfig.StatusPort != 0 {
+		http.HandleFunc("/_status", statusHandler(successMainUpdate, successHookUpdate, errMainUpdate, errHookUpdate))
+		go http.ListenAndServe(fmt.Sprintf(":%d", preparerConfig.StatusPort), nil)
+	} else {
+		// Discard everything from these channel!
+		go func() {
+			for {
+				select {
+				case <-successMainUpdate:
+				case <-successHookUpdate:
+				case <-errMainUpdate:
+				case <-errHookUpdate:
+				}
+			}
+		}()
+	}
 
 	waitForTermination(logger, quitMainUpdate, quitHookUpdate)
 
 	logger.NoFields().Infoln("Terminating")
 }
 
+func watchForErrors(successes <-chan struct{}, errs <-chan error, consecutive *int, lastError *error) {
+	for {
+		select {
+		case err := <-errs:
+			*lastError = err
+			*consecutive++
+		case <-successes:
+			*consecutive = 0
+		}
+	}
+}
+
+// if the number of consecutive errors >= tooManyErrors, the status is critical.
+const tooManyErrors = 10
+
+func statusHandler(mainSuccesses, hookSuccesses <-chan struct{}, mainErrors, hookErrors <-chan error) http.HandlerFunc {
+	consecutiveMainErrors := 0
```

Comment:
There are data races accessing all four of these variables. Add appropriate synchronization.